binary_search_tree/tree.py

```diff
@@ -14,46 +14,110 @@ class Tree:
     def __init__(self):
         self.root = None
 
-    # Time Complexity: 
-    # Space Complexity: 
+    #----------------------- ADD ----------------------
+    #from from class video 👇🏼
+    def add_helper(self, current, key, value):
+        if current == None:
+            return TreeNode(key, value)
+        elif current.key >= key:
+            current.left = self.add_helper(current.left, key, value)
+        else: 
+            current.right = self.add_helper(current.right, key, value)
+        return current 
+    #from from class video 👆🏼
+
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
     def add(self, key, value = None):
-        pass
 
-    # Time Complexity: 
-    # Space Complexity: 
-    def find(self, key):
-        pass
+        if self.root is None:
+            self.root = TreeNode(key, value)
+            return
+        self.add_helper(self.root, key, value)
 
-    # Time Complexity: 
-    # Space Complexity: 
+    #----------------------- FIND ----------------------
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n)
+    def find_helper(self, current, key):
+        if current.key == key:
+            return current.value
+        if current.key >= key:
+            if not current.left:
+                return None
+            return self.find_helper(current.left, key)
+        if not current.right:
+            return None
+        return self.find_helper(current.right, key)
+    
+    def find(self, key):
+        if self.root == None:
+            return None
+        return self.find_helper(self.root, key)
+        
+    #----------------------- IN ORDER ----------------------
+    # Time Complexity: O(log n)
+    # Space Complexity: O(log n) 
```

Comment:
Note that unlike add and find where we only traverse down one branch of the tree, that when we do the depth-first traversals, we still do hit every node in the tree making the time and space complexity both O(n).

Reply:
This applies to `inorder`, `preorder`, and `postorder`.